File: modules/downloader.py
# modules/downloader.py
# Handles saving downloaded images and videos to disk with unique filenames.

import os
import shutil
import logging
from pathlib import Path
from config import IMAGES_DIR, VIDEOS_DIR

logger = logging.getLogger("bot")


class Downloader:
    """
    Saves files that Playwright downloads to the correct output folder.
    Also handles copying an already-downloaded temp file to its final path.
    """

    @staticmethod
    def save_image(download, prompt_id: int) -> str:
        """
        Accept a Playwright Download object, save it to images/<id>.png.
        Returns the saved file path.
        """
        filename = f"{prompt_id}.png"
        dest     = os.path.join(IMAGES_DIR, filename)
        download.save_as(dest)
        logger.info(f"[ID:{prompt_id}] Image saved  ->  {dest}")
        return dest

    @staticmethod
    def save_video(download, prompt_id: int) -> str:
        """
        Accept a Playwright Download object, save it to videos/<id>.mp4.
        Returns the saved file path.
        """
        filename = f"{prompt_id}.mp4"
        dest     = os.path.join(VIDEOS_DIR, filename)
        download.save_as(dest)
        logger.info(f"[ID:{prompt_id}] Video saved  ->  {dest}")
        return dest

    @staticmethod
    def copy_file(src: str, dest: str) -> str:
        """
        Copy an existing file from src to dest.
        Used when the browser saves to a temp location.
        """
        os.makedirs(os.path.dirname(dest), exist_ok=True)
        shutil.copy2(src, dest)
        logger.info(f"File copied  {src}  ->  {dest}")
        return dest

    @staticmethod
    def get_image_path(prompt_id: int) -> str:
        return os.path.join(IMAGES_DIR, f"{prompt_id}.png")

    @staticmethod
    def get_video_path(prompt_id: int) -> str:
        return os.path.join(VIDEOS_DIR, f"{prompt_id}.mp4")

    @staticmethod
    def image_exists(prompt_id: int) -> bool:
        return Path(Downloader.get_image_path(prompt_id)).exists()

    @staticmethod
    def video_exists(prompt_id: int) -> bool:
        return Path(Downloader.get_video_path(prompt_id)).exists()

File: src/downloader.ts
// src/downloader.ts
// Handles saving downloaded images and videos to disk.

import path from 'path';
import fs from 'fs';
import { IMAGES_DIR, VIDEOS_DIR } from './config';
import logger from './logger';
import { Download } from 'playwright';

export class Downloader {
  static getImagePath(id: number): string {
    return path.join(IMAGES_DIR, `${id}.png`);
  }

  static getVideoPath(id: number): string {
    return path.join(VIDEOS_DIR, `${id}.mp4`);
  }

  static imageExists(id: number): boolean {
    return fs.existsSync(Downloader.getImagePath(id));
  }

  static videoExists(id: number): boolean {
    return fs.existsSync(Downloader.getVideoPath(id));
  }

  static async saveImage(download: Download, id: number): Promise<string> {
    const dest = Downloader.getImagePath(id);
    await download.saveAs(dest);
    logger.info(`[ID:${id}] Image saved -> ${dest}`);
    return dest;
  }

  static async saveVideo(download: Download, id: number): Promise<string> {
    const dest = Downloader.getVideoPath(id);
    await download.saveAs(dest);
    logger.info(`[ID:${id}] Video saved -> ${dest}`);
    return dest;
  }

  static copyFile(src: string, dest: string): string {
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    fs.copyFileSync(src, dest);
    logger.info(`File copied: ${src} -> ${dest}`);
    return dest;
  }
}
